src/EFCore.PG/Storage/Internal/Mapping/NpgsqlArrayTypeMapping.cs

```diff
@@ -1,67 +1,49 @@
     /// See: https://www.postgresql.org/docs/current/static/arrays.html
     /// </remarks>
-    public class NpgsqlArrayTypeMapping : RelationalTypeMapping
+    public abstract class NpgsqlArrayTypeMapping : RelationalTypeMapping
     {
-        // ReSharper disable once MemberCanBePrivate.Global
         /// <summary>
         /// The relational type mapping used to initialize the array mapping.
         /// </summary>
         public RelationalTypeMapping ElementMapping { get; }
 
-        /// <summary>
-        /// Creates the default array mapping (i.e. for the single-dimensional CLR array type)
-        /// </summary>
-        /// <param name="storeType">The database type to map.</param>
-        /// <param name="elementMapping">The element type mapping.</param>
-        public NpgsqlArrayTypeMapping(string storeType, RelationalTypeMapping elementMapping)
-            : this(storeType, elementMapping, elementMapping.ClrType.MakeArrayType()) {}
-
-        /// <summary>
-        /// Creates the default array mapping (i.e. for the single-dimensional CLR array type)
-        /// </summary>
-        /// <param name="elementMapping">The element type mapping.</param>
-        /// <param name="arrayType">The array type to map.</param>
-        public NpgsqlArrayTypeMapping(RelationalTypeMapping elementMapping, Type arrayType)
-            : this(elementMapping.StoreType + "[]", elementMapping, arrayType) {}
-
-        NpgsqlArrayTypeMapping(string storeType, RelationalTypeMapping elementMapping, Type arrayType)
-            : this(new RelationalTypeMappingParameters(
-                new CoreTypeMappingParameters(arrayType, null, CreateComparer(elementMapping, arrayType)), storeType
-            ), elementMapping) {}
-
         protected NpgsqlArrayTypeMapping(RelationalTypeMappingParameters parameters, RelationalTypeMapping elementMapping)
             : base(parameters)
             => ElementMapping = elementMapping;
 
-        protected override RelationalTypeMapping Clone(RelationalTypeMappingParameters parameters)
-            => new NpgsqlArrayTypeMapping(parameters, ElementMapping);
-
+        // The array-to-array mapping needs to know how to generate an SQL literal for a List<>, and
+        // the list-to-array mapping needs to know how to generate an SQL literal for an array.
+        // This is because in cases such as ctx.SomeListColumn.SequenceEquals(new[] { 1, 2, 3}), the list mapping
+        // from the left side gets applied to the right side.
         protected override string GenerateNonNullSqlLiteral(object value)
         {
-            var arr = (Array)value;
+            var type = value.GetType();
 
-            if (arr.Rank != 1)
+            if (!type.IsArray && !type.IsGenericList())
+                throw new ArgumentException("Parameter must be an array or List<>", nameof(value));
+            if (type.IsArray && ((Array)value).Rank != 1)
```

Comment:
The runtime woon't eliminate type check on casting, so:
```suggestion
            if (value is Array array && array.Rank != 1)
```

Otherwise, you should use `Unsafe.As<>`, but it's ugly as you know.

Reply:
Note that this is totally perf-insensitive code (one-time generation of SQL literals), so stuff like type check elimination isn't very relevant... But your suggestions looks better.